core/src/main/java/io/cucumber/core/filter/TagPredicate.java

```diff
@@ -2,40 +2,28 @@
 
 import io.cucumber.core.gherkin.Pickle;
 import io.cucumber.tagexpressions.Expression;
-import io.cucumber.tagexpressions.TagExpressionException;
-import io.cucumber.tagexpressions.TagExpressionParser;
 
-import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Predicate;
-import java.util.stream.Collectors;
 
 import static java.util.Collections.emptyList;
-import static java.util.Collections.singletonList;
 
 final class TagPredicate implements Predicate<Pickle> {
 
     private final List<Expression> expressions;
 
-    TagPredicate(String tagExpression) {
-        this(tagExpression.isEmpty() ? emptyList() : singletonList(tagExpression));
+    TagPredicate(Expression tagExpression) {
+        this(Optional.ofNullable(tagExpression)
+                .map(Collections::singletonList)
+                .orElse(emptyList()));
```

Comment:
Generally it is preferable to avoid nullable parameters.  The original `tagExpression` was already non-null because invoking  `tagExpression.isEmpty()` so you may be able to reduce this to

 ```
this(singletonList(requireNonNull(tagExpression))`;
```

Reply:
Convention wise, using `Optional` to do a null check is not prefered. A ternary is usually more readable.

Reply:
Ok, no problem.  I tend to avoid ternary as I find them harder to read myself.